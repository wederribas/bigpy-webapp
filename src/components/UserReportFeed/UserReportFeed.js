import React from 'react'
import PropTypes from 'prop-types'
import { Query } from 'react-apollo'
import gql from 'graphql-tag'
import {
  Button,
  Paper,
  Typography,
  withStyles,
  List,
  ListItem,
  ListItemText,
  Divider
} from '@material-ui/core'
import ErrorMessage from '../ErrorMessage/ErrorMessage'

const styles = theme => ({
  root: {
    ...theme.mixins.gutters(),
    paddingTop: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2
  },
  feed: {
    marginTop: theme.spacing.unit * 2
  },
  feedListItem: {
    display: 'block'
  }
})

export const USER_REPORT_FEED_QUERY = gql`
  query UserReportFeedQuery($companyName: String!, $cursor: String) {
    allCustomerReports(after: $cursor, first: 5, companyName: $companyName) {
      pageInfo {
        endCursor
        hasNextPage
      }
      edges {
        node {
          Id
          companyName
          city
          state
          userReport
          userRating
          userFeedback
          date
        }
      }
    }
  }
`

function LoadMoreButton(props) {
  const { classes } = props
  return (
    <Button
      variant="contained"
      color="primary"
      className={classes.button}
      onClick={() => props.loadMore()}
    >
      Load More
    </Button>
  )
}

LoadMoreButton.propTypes = {
  classes: PropTypes.object.isRequired,
  loadMore: PropTypes.func.isRequired
}

function ReportsFeed(props) {
  const { classes } = props
  const totalReports = props.reports.length
  return (
    <Paper elevation={1} className={`${classes.root} ${props.classes.feed}`}>
      <Typography variant="h4" component="h2">
        Users Reports
      </Typography>
      <List data-testid="reports-list">
        {props.reports.map((report, index) => {
          const date = new Date(report.node.date)
          return (
            <div key={report.node.Id}>
              <ListItem className={classes.feedListItem}>
                <ListItemText
                  primary={`${report.node.city} - ${report.node.state}`}
                  secondary={date.toDateString()}
                />
                <Typography component="p">{report.node.userReport}</Typography>
                <Typography component="p">
                  {report.node.userFeedback}
                </Typography>
                <Typography component="p">{report.node.userRating}</Typography>
              </ListItem>
              {totalReports === index + 1 || <Divider />}
            </div>
          )
        })}
      </List>
      <LoadMoreButton loadMore={props.loadMore} classes={props.classes} />
    </Paper>
  )
}

ReportsFeed.propTypes = {
  classes: PropTypes.object.isRequired,
  reports: PropTypes.array.isRequired,
  loadMore: PropTypes.func.isRequired
}

function UserReportFeed(props) {
  return (
    <Query
      query={USER_REPORT_FEED_QUERY}
      variables={{ companyName: props.companyName }}
    >
      {({ loading, error, data, fetchMore }) => {
        if (loading) return null

        if (error) return <ErrorMessage />

        const paginationCursor = data.allCustomerReports.pageInfo.endCursor
        const reports = data.allCustomerReports.edges

        const loadMore = () =>
          fetchMore({
            variables: {
              cursor: paginationCursor
            },
            updateQuery: (previousResult, { fetchMoreResult }) => {
              const newEdges = fetchMoreResult.allCustomerReports.edges
              const pageInfo = fetchMoreResult.allCustomerReports.pageInfo

              return newEdges.length
                ? {
                    allCustomerReports: {
                      __typename: previousResult.allCustomerReports.__typename,
                      edges: [
                        ...previousResult.allCustomerReports.edges,
                        ...newEdges
                      ],
                      pageInfo
                    }
                  }
                : previousResult
            }
          })

        return (
          <div data-testid="feed">
            <ReportsFeed
              reports={reports}
              classes={props.classes}
              loadMore={loadMore}
            />
          </div>
        )
      }}
    </Query>
  )
}

UserReportFeed.propTypes = {
  companyName: PropTypes.string.isRequired,
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(UserReportFeed)
