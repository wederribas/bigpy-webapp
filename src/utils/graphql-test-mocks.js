import { getCompaniesQuery } from '../components/Search/Search'

export const mocks = [
  {
    request: {
      query: getCompaniesQuery
    },
    result: {
      data: {
        allCompaniesNames: {
          edges: [
            { node: { Id: 'Company 0' } },
            { node: { Id: 'Company 1' } },
            { node: { Id: 'Company 2' } }
          ]
        }
      }
    }
  }
]
