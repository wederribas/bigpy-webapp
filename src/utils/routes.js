import Search from '../components/Search/Search'
import Top10 from '../components/Top10/Top10'
import Insights from '../components/Insights/Insights'
import About from '../components/About/About'
import Dashboard from '../components/Dashboard/Dashboard'

export const routes = [
  {
    path: '/',
    component: Search,
    exact: true
  },
  {
    path: '/dashboard/:companyName',
    component: Dashboard
  },
  {
    path: '/top10',
    component: Top10
  },
  {
    path: '/insights',
    component: Insights
  },
  {
    path: '/about',
    component: About
  }
]
