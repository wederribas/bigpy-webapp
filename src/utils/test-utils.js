import React from 'react'
import { render } from 'react-testing-library'
import { MuiThemeProvider, createMuiTheme } from '@material-ui/core/styles'

const theme = createMuiTheme({
  typography: {
    useNextVariants: true
  }
})

const customRender = node => {
  return render(<MuiThemeProvider theme={theme}>{node}</MuiThemeProvider>)
}

export * from 'react-testing-library'

export { customRender as render }
