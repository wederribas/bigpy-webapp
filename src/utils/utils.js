// @flow

/**
 * Deburr special Latin letters to basic Latin letters and removing
 * combining diacritical marks.
 */
export function deburr(string: string): string {
  return string.normalize('NFD').replace(/[\u0300-\u036f]/g, '')
}
