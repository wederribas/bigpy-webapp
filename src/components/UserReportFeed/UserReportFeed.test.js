import React from 'react'
import { MockedProvider } from 'react-apollo/test-utils'
import waait from 'waait'
import { render } from '../../utils/test-utils'
import UserReportFeed, { USER_REPORT_FEED_QUERY } from './UserReportFeed'

const queryData = {
  companyName: 'Company 0',
  userReport: 'This is my user report',
  city: 'Narnia City',
  state: 'BA',
  userRating: 5,
  userFeedback: 'This is the user feedback',
  date: new Date('1536894000000')
}

const userReportMock = {
  request: {
    query: USER_REPORT_FEED_QUERY,
    variables: {
      companyName: queryData.companyName
    }
  },
  result: {
    data: {
      allCustomerReports: {
        pageInfo: {
          endCursor: 'stringXYZ',
          hasNextPage: true
        },
        edges: [
          {
            node: {
              Id: 'id@123',
              companyName: queryData.companyName,
              userReport: queryData.userReport,
              city: queryData.city,
              state: queryData.state,
              userRating: queryData.userRating,
              userFeedback: queryData.userFeedback,
              date: queryData.date.toDateString()
            }
          }
        ]
      }
    }
  }
}

test('<UserReportFeed /> component renders', async () => {
  const { container, getByTestId, rerender } = render(
    <MockedProvider mocks={[userReportMock]} addTypename={false}>
      <UserReportFeed companyName={queryData.companyName} />
    </MockedProvider>
  )

  expect(container.firstChild).toMatchSnapshot()

  await waait(200)

  expect(getByTestId('feed')).toBeTruthy()
  expect(getByTestId('reports-list').children).toHaveLength(1)

  // Re-render the component with error
  delete userReportMock.result
  userReportMock.error = { error: new Error('Oh snap!') }
  rerender(
    <MockedProvider mocks={[userReportMock]} addTypename={false}>
      <UserReportFeed companyName={queryData.companyName} />
    </MockedProvider>
  )

  await waait(200)

  expect(getByTestId('graphql-error')).toBeTruthy()
})
