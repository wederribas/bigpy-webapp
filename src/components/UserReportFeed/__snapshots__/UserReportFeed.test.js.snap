// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserReportFeed /> component renders 1`] = `null`;
